lib/Language/Slicer/Absyn.hs

```diff
@@ -105,8 +108,28 @@ data Type = IntTy | BoolTy | UnitTy | StringTy
           | TyVar TyVar
           -- References
           | RefTy Type
+          -- Exception type
+          | ExnTy
           -- Trace types
           | TraceTy Type
-            deriving (Show, Eq, Ord, Generic, NFData)
+            deriving (Show, Generic, NFData)
+
+-- We need a hand-written instance of Eq, because ExnTy is equal to every type
+instance Eq Type where
+    ExnTy        == _              = True
```

Comment:
Yes, I'm happy with postponing implementing type inference etc.  One approach to avoiding this could be to tag "raise" with the return type it should have, similar to "hole".